[Forms] Use button instead of push-button appearance for input[type=submit|reset|button]

nt1m opened the pull request:
#### 1a0105921defca0770820813cdec1e805036c441
<pre>
[Forms] Use button instead of push-button appearance for input[type=submit|reset|button]
<a href="https://bugs.webkit.org/show_bug.cgi?id=238803">https://bugs.webkit.org/show_bug.cgi?id=238803</a>
rdar://91624840

Reviewed by Aditya Keerthi.

- Fixes bug reported by GOV.UK about font-size not applying on input[type=file]: <a href="https://bugs.webkit.org/show_bug.cgi?id=224746">https://bugs.webkit.org/show_bug.cgi?id=224746</a>
- Fixes input[type=button|reset|submit] not supporting multi-line values: <a href="https://bugs.webkit.org/show_bug.cgi?id=190521">https://bugs.webkit.org/show_bug.cgi?id=190521</a>
- Fixes input[type=button|reset|submit] not honoring font-size/padding/height/etc. properly like other browsers

* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::autoAppearanceForElement const):

Use StyleAppearance::Button instead of StyleAppearance::PushButton

* LayoutTests/fast/css/button-height-expected.txt:
* LayoutTests/fast/css/button-height.html:

Adjust test now that input[type=button] honors height on macOS.

* LayoutTests/platform/ios/css2.1/20110323/replaced-elements-001-expected.txt:
* LayoutTests/platform/ios/css3/flexbox/flexitem-expected.txt:
* LayoutTests/platform/ios/fast/forms/button-style-color-expected.txt:
* LayoutTests/platform/mac-bigsur/fast/css/continuationCrash-expected.txt:
* LayoutTests/platform/mac-bigsur/fast/forms/button-sizes-expected.txt:
* LayoutTests/platform/mac-bigsur/fast/forms/control-restrict-line-height-expected.txt:
* LayoutTests/platform/mac-bigsur/tables/mozilla/bugs/bug2479-3-expected.txt:
* LayoutTests/platform/mac-bigsur/tables/mozilla/bugs/bug26178-expected.txt:
* LayoutTests/platform/mac-monterey/fast/forms/button-sizes-expected.txt:
* LayoutTests/platform/mac-ventura/fast/forms/button-sizes-expected.txt:
* LayoutTests/platform/mac-ventura/fast/forms/input-button-sizes-expected.txt: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/foreign-object-skew-expected.txt:
* LayoutTests/platform/mac/css3/flexbox/button-expected.txt:
* LayoutTests/platform/mac/editing/selection/3690703-2-expected.txt:
* LayoutTests/platform/mac/editing/selection/3690703-expected.txt:
* LayoutTests/platform/mac/editing/selection/3690719-expected.txt:
* LayoutTests/platform/mac/editing/selection/4397952-expected.txt:
* LayoutTests/platform/mac/editing/selection/5240265-expected.txt:
* LayoutTests/platform/mac/editing/selection/selection-button-text-expected.txt:
* LayoutTests/platform/mac/fast/block/float/float-avoidance-expected.txt:
* LayoutTests/platform/mac/fast/css/continuationCrash-expected.txt:
* LayoutTests/platform/mac/fast/css/input-search-padding-expected.txt:
* LayoutTests/platform/mac/fast/css/margin-top-bottom-dynamic-expected.txt:
* LayoutTests/platform/mac/fast/css/rtl-ordering-expected.txt:
* LayoutTests/platform/mac/fast/dom/HTMLTextAreaElement/reset-textarea-expected.txt:
* LayoutTests/platform/mac/fast/forms/001-expected.txt:
* LayoutTests/platform/mac/fast/forms/basic-buttons-expected.txt:
* LayoutTests/platform/mac/fast/forms/blankbuttons-expected.txt:
* LayoutTests/platform/mac/fast/forms/box-shadow-override-expected.txt:
* LayoutTests/platform/mac/fast/forms/button-positioned-expected.txt:
* LayoutTests/platform/mac/fast/forms/button-style-color-expected.txt:
* LayoutTests/platform/mac/fast/forms/button-table-styles-expected.txt:
* LayoutTests/platform/mac/fast/forms/button-text-transform-expected.txt:
* LayoutTests/platform/mac/fast/forms/control-restrict-line-height-expected.txt:
* LayoutTests/platform/mac/fast/forms/file/file-input-direction-expected.txt:
* LayoutTests/platform/mac/fast/forms/file/file-input-disabled-expected.txt:
* LayoutTests/platform/mac/fast/forms/file/input-file-re-render-expected.txt:
* LayoutTests/platform/mac/fast/forms/form-element-geometry-expected.txt:
* LayoutTests/platform/mac/fast/forms/formmove3-expected.txt:
* LayoutTests/platform/mac/fast/forms/input-appearance-height-expected.txt:
* LayoutTests/platform/mac/fast/forms/input-button-sizes-expected.txt:
* LayoutTests/platform/mac/fast/forms/input-value-expected.txt:
* LayoutTests/platform/mac/fast/forms/targeted-frame-submission-expected.txt:
* LayoutTests/platform/mac/fast/html/details-replace-summary-child-expected.txt:
* LayoutTests/platform/mac/fast/html/details-replace-text-expected.txt:
* LayoutTests/platform/mac/fast/overflow/scroll-nested-positioned-layer-in-overflow-expected.txt:
* LayoutTests/platform/mac/fast/overflow/scrollRevealButton-expected.txt:
* LayoutTests/platform/mac/fast/replaced/replaced-breaking-expected.txt:
* LayoutTests/platform/mac/fast/replaced/width100percent-button-expected.txt:
* LayoutTests/platform/mac/fast/text/international/hindi-spacing-expected.txt:
* LayoutTests/platform/mac/fast/text/textIteratorNilRenderer-expected.txt:
* LayoutTests/platform/mac/http/tests/navigation/javascriptlink-frames-expected.txt:
* LayoutTests/platform/mac/svg/custom/foreign-object-skew-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug1188-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug1318-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug138725-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug18359-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug2479-2-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug2479-3-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug2479-4-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug26178-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug28928-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug33855-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug39209-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug4429-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug46368-1-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug46368-2-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug51037-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug51727-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug52505-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug52506-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug60749-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug7342-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug92647-2-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/collapsing_borders/bug41262-4-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/dom/tableDom-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/other/move_row-expected.txt:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/bugs/bug1725-expected.txt:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/bugs/bug2479-5-expected.txt:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/bugs/bug58402-2-expected.txt:

Rebaseline a bunch of tests.

Canonical link: <a href="https://commits.webkit.org/258754@main">https://commits.webkit.org/258754@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e46ed4894e74b16706c05c78dca2cedfae49f55a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102848 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11966 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/35874 "Build was cancelled. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings); Failed to upload built product") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112102 "Built successfully") | [💥 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172322 "An unexpected error occured. Recent messages:Pull request contains relevant changes; Failed to print configuration") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12983 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2875 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95096 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109772 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108622 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/35874 "Build was cancelled. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings); Failed to upload built product") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37610 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/35874 "Build was cancelled. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings); Failed to upload built product") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79344 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5419 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/35874 "Build was cancelled. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings); Failed to upload built product") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5567 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2573 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11586 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/35874 "Build was cancelled. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings); Failed to upload built product") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6017 "Failed to push to pull request branch") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7312 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3201 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -373,7 +373,7 @@ StyleAppearance RenderTheme::autoAppearanceForElement(RenderStyle& style, const
         auto& input = downcast<HTMLInputElement>(element.get());
 
         if (input.isTextButton() || input.isUploadButton())
-            return StyleAppearance::PushButton;
```
Does this mean we can remove `StyleAppearance::PushButton` entirely?

nt1m replied:
Yes, I'd prefer doing so in another bug, once we're sure we don't want to use it (in internal apps or such), so we can add an internal value later on if necessary.

pxlcoder commented on LayoutTests/platform/ios/fast/forms/button-style-color-expected.txt:
```diff
@@ -39,14 +39,14 @@ layer at (0,0) size 800x600
         text run at (525,0) width 5: " "
-      RenderButton {INPUT} at (529,1) size 85x20 [color=#007AFF] [bgcolor=#008000] [border: (1px solid #FFFFFF)]
-        RenderBlock (anonymous) at (12,3) size 61x14
+      RenderButton {INPUT} at (529,3) size 85x16 [color=#007AFF] [bgcolor=#008000] [border: (1px solid #FFFFFF)]
```
What's the reason behind the height decrease on iOS?

nt1m replied:
Not sure, but it matches the `<button>` above now.

pxlcoder commented on LayoutTests/platform/mac-bigsur/tables/mozilla/bugs/bug2479-3-expected.txt:
```diff
@@ -85,7 +85,7 @@ layer at (0,0) size 785x765
           RenderText {#text} at (548,17) size 17x16
             text run at (548,17) width 17: " "
-          RenderButton {INPUT} at (564,17) size 54x18 [color=#000000D8] [bgcolor=#C0C0C0]
+          RenderButton {INPUT} at (564,17) size 54x18 [color=#000000D8] [bgcolor=#C0C0C0] [border: none (2px outset #C0C0C0) none (2px outset #C0C0C0)]
```
Where does the border come from here? Given that the size hasn't changed, I don't think there should be rendering differences?

nt1m replied:
There's some 2px spacing on the left/right side that now appears (top/bottom is `none` like the dump shows), similar to the one on `<button>` currently, but it's not visible anyway since the native appearance.

nt1m committed:
[Forms] Use button instead of push-button appearance for input[type=submit|reset|button]
https://bugs.webkit.org/show_bug.cgi?id=238803
rdar://91624840

Reviewed by Aditya Keerthi.

- Fixes bug reported by GOV.UK about font-size not applying on input[type=file]: https://bugs.webkit.org/show_bug.cgi?id=224746
- Fixes input[type=button|reset|submit] not supporting multi-line values: https://bugs.webkit.org/show_bug.cgi?id=190521
- Fixes input[type=button|reset|submit] not honoring font-size/padding/height/etc. properly like other browsers

* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::autoAppearanceForElement const):

Use StyleAppearance::Button instead of StyleAppearance::PushButton

* LayoutTests/fast/css/button-height-expected.txt:
* LayoutTests/fast/css/button-height.html:

Adjust test now that input[type=button] honors height on macOS.

* LayoutTests/platform/ios/css2.1/20110323/replaced-elements-001-expected.txt:
* LayoutTests/platform/ios/css3/flexbox/flexitem-expected.txt:
* LayoutTests/platform/ios/fast/forms/button-style-color-expected.txt:
* LayoutTests/platform/mac-bigsur/fast/css/continuationCrash-expected.txt:
* LayoutTests/platform/mac-bigsur/fast/forms/button-sizes-expected.txt:
* LayoutTests/platform/mac-bigsur/fast/forms/control-restrict-line-height-expected.txt:
* LayoutTests/platform/mac-bigsur/tables/mozilla/bugs/bug2479-3-expected.txt:
* LayoutTests/platform/mac-bigsur/tables/mozilla/bugs/bug26178-expected.txt:
* LayoutTests/platform/mac-monterey/fast/forms/button-sizes-expected.txt:
* LayoutTests/platform/mac-ventura/fast/forms/button-sizes-expected.txt:
* LayoutTests/platform/mac-ventura/fast/forms/input-button-sizes-expected.txt: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/foreign-object-skew-expected.txt:
* LayoutTests/platform/mac/css3/flexbox/button-expected.txt:
* LayoutTests/platform/mac/editing/selection/3690703-2-expected.txt:
* LayoutTests/platform/mac/editing/selection/3690703-expected.txt:
* LayoutTests/platform/mac/editing/selection/3690719-expected.txt:
* LayoutTests/platform/mac/editing/selection/4397952-expected.txt:
* LayoutTests/platform/mac/editing/selection/5240265-expected.txt:
* LayoutTests/platform/mac/editing/selection/selection-button-text-expected.txt:
* LayoutTests/platform/mac/fast/block/float/float-avoidance-expected.txt:
* LayoutTests/platform/mac/fast/css/continuationCrash-expected.txt:
* LayoutTests/platform/mac/fast/css/input-search-padding-expected.txt:
* LayoutTests/platform/mac/fast/css/margin-top-bottom-dynamic-expected.txt:
* LayoutTests/platform/mac/fast/css/rtl-ordering-expected.txt:
* LayoutTests/platform/mac/fast/dom/HTMLTextAreaElement/reset-textarea-expected.txt:
* LayoutTests/platform/mac/fast/forms/001-expected.txt:
* LayoutTests/platform/mac/fast/forms/basic-buttons-expected.txt:
* LayoutTests/platform/mac/fast/forms/blankbuttons-expected.txt:
* LayoutTests/platform/mac/fast/forms/box-shadow-override-expected.txt:
* LayoutTests/platform/mac/fast/forms/button-positioned-expected.txt:
* LayoutTests/platform/mac/fast/forms/button-style-color-expected.txt:
* LayoutTests/platform/mac/fast/forms/button-table-styles-expected.txt:
* LayoutTests/platform/mac/fast/forms/button-text-transform-expected.txt:
* LayoutTests/platform/mac/fast/forms/control-restrict-line-height-expected.txt:
* LayoutTests/platform/mac/fast/forms/file/file-input-direction-expected.txt:
* LayoutTests/platform/mac/fast/forms/file/file-input-disabled-expected.txt:
* LayoutTests/platform/mac/fast/forms/file/input-file-re-render-expected.txt:
* LayoutTests/platform/mac/fast/forms/form-element-geometry-expected.txt:
* LayoutTests/platform/mac/fast/forms/formmove3-expected.txt:
* LayoutTests/platform/mac/fast/forms/input-appearance-height-expected.txt:
* LayoutTests/platform/mac/fast/forms/input-button-sizes-expected.txt:
* LayoutTests/platform/mac/fast/forms/input-value-expected.txt:
* LayoutTests/platform/mac/fast/forms/targeted-frame-submission-expected.txt:
* LayoutTests/platform/mac/fast/html/details-replace-summary-child-expected.txt:
* LayoutTests/platform/mac/fast/html/details-replace-text-expected.txt:
* LayoutTests/platform/mac/fast/overflow/scroll-nested-positioned-layer-in-overflow-expected.txt:
* LayoutTests/platform/mac/fast/overflow/scrollRevealButton-expected.txt:
* LayoutTests/platform/mac/fast/replaced/replaced-breaking-expected.txt:
* LayoutTests/platform/mac/fast/replaced/width100percent-button-expected.txt:
* LayoutTests/platform/mac/fast/text/international/hindi-spacing-expected.txt:
* LayoutTests/platform/mac/fast/text/textIteratorNilRenderer-expected.txt:
* LayoutTests/platform/mac/http/tests/navigation/javascriptlink-frames-expected.txt:
* LayoutTests/platform/mac/svg/custom/foreign-object-skew-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug1188-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug1318-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug138725-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug18359-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug2479-2-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug2479-3-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug2479-4-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug26178-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug28928-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug33855-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug39209-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug4429-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug46368-1-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug46368-2-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug51037-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug51727-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug52505-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug52506-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug60749-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug7342-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug92647-2-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/collapsing_borders/bug41262-4-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/dom/tableDom-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/other/move_row-expected.txt:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/bugs/bug1725-expected.txt:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/bugs/bug2479-5-expected.txt:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/bugs/bug58402-2-expected.txt:

Rebaseline a bunch of tests.

Canonical link: https://commits.webkit.org/258754@main
